File: examples/titan_elixir/apps/titan/lib/titan_ecron_logger.ex
defmodule TitanEcronLogger do
  @moduledoc false

  require Logger

  @events [
    [:ecron, :activate],
    [:ecron, :deactivate],
    [:ecron, :delete],
    [:ecron, :success],
    [:ecron, :failure],
    [:ecron, :global, :up],
    [:ecron, :global, :down],
  ]

  def attach(), do: :telemetry.attach_many(:ecron_metrics, @events, &handle_event/4, :undefined)
  def detach(), do: :telemetry.detach(:ecron_metrics)

  def handle_event([:ecron, :activate], %{action_ms: time}, %{name: name, mfa: mfa}, _config) do
    Logger.info("Activate ecron #{inspect name} #{inspect mfa}  at #{inspect time}")
  end

  def handle_event([:ecron, :deactivate], %{action_ms: time}, %{name: name}, _config) do
    Logger.info("Deactivate ecron #{inspect name} at #{inspect time}")
  end

  def handle_event([:ecron, :delete], %{action_ms: time}, %{name: name}, _config) do
    Logger.info("Delete ecron #{inspect name} at #{inspect time}")
  end

  def handle_event([:ecron, :success], %{run_microsecond: ms, run_result: res}, %{name: name, mfa: mfa}, _config) do
    Logger.info(
      "Ecron #{inspect name}(#{inspect mfa}) completed in #{inspect ms} microsecond. result is #{inspect res}"
    )
  end

  def handle_event([:ecron, :failure], %{run_microsecond: ms, run_result: res}, %{name: name, mfa: mfa}, _config) do
    Logger.error("Ecron #{inspect name}(#{inspect mfa}) CRASH in #{inspect ms} microsecond. result is #{inspect res}")
  end

  def handle_event([:ecron, :global, :up], measurements, %{self: node}, _config) do
    %{action_ms: ms, quorum_size: quorum_size, good_nodes: good_nodes, bad_nodes: bad_nodes} = measurements
    Logger.info(
      "Ecron global UP on #{inspect node} at #{inspect ms} ms, quorum_size is #{quorum_size} good_nodes is #{
        inspect good_nodes
      }} bad_nodes is #{inspect bad_nodes}."
    )
  end

  def handle_event([:ecron, :global, :down], measurements, %{self: node}, _config) do
    %{action_ms: ms, quorum_size: quorum_size, good_nodes: good_nodes, bad_nodes: bad_nodes} = measurements
    Logger.info(
      "Ecron global Down on #{inspect node} at #{inspect ms} ms, quorum_size is #{quorum_size} good_nodes is #{
        inspect good_nodes
      }} bad_nodes is #{inspect bad_nodes}."
    )
  end

end

File: examples/titan_erlang/apps/titan/src/titan_ecron_logger.erl
-module(titan_ecron_logger).

-include_lib("kernel/include/logger.hrl").
-define(Success, [ecron, success]).
-define(Failure, [ecron, failure]).
-define(Activate, [ecron, activate]).
-define(Deactivate, [ecron, deactivate]).
-define(Delete, [ecron, delete]).
-define(GlobalUp, [ecron, global, up]).
-define(GlobalDown, [ecron, global, down]).

-define(Events, [?Success, ?Failure, ?Activate, ?Deactivate, ?Delete, ?GlobalUp, ?GlobalDown]).
%% API
-export([attach/0, detach/0]).
-define(TELEMETRY_HANDLE, ecron_telemetry_metrics).

attach() ->
    telemetry:attach_many(?TELEMETRY_HANDLE, ?Events, fun handle_event/4, undefined).

detach() ->
    telemetry:detach(?TELEMETRY_HANDLE).

handle_event(?Activate, #{action_ms := Time}, #{name := Name, mfa := MFA}, _Config) ->
    ?LOG_INFO("EcronJob(~p)-~p activated at ~p .", [Name, MFA, Time]);
handle_event(?Deactivate, #{action_ms := Time}, #{name := Name}, _Config) ->
    ?LOG_INFO("EcronJob(~p) deactivated at ~p .", [Name, Time]);
handle_event(?Delete, #{action_ms := Time}, #{name := Name}, _Config) ->
    ?LOG_INFO("EcronJob(~p) deleted at ~p .", [Name, Time]);
handle_event(?Success, #{run_microsecond := Ms, run_result := Res},
    #{name := Name, mfa := {M, F, A}}, _Config) ->
    ?LOG_INFO("EcronJob(~p) completed in ~p microsecond.~p=apply(~p,~p,~p)", [Name, Ms, Res, M, F, A]);
handle_event(?Success, #{run_microsecond := Ms, run_result := Res},
    #{name := Name, mfa := {F, A}}, _Config) ->
    ?LOG_INFO("EcronJob(~p) completed in ~p microsecond.~p=apply(~p,~p)", [Name, Ms, Res, F, A]);
handle_event(?Failure, #{run_microsecond := Ms, run_result := {Error, Reason, Stack}},
    #{name := Name, mfa := MFA}, _Config) ->
    ?LOG_ERROR("EcronJob(~p)-~p CRASH in ~p microsecond. {Error, Reason}: {~p, ~p}. Stack:~p",
        [Name, MFA, Ms, Error, Reason, Stack]);
handle_event(?GlobalUp, #{action_ms := Time, quorum_size := QuorumSize,
    good_nodes := GoodNodes, bad_nodes := BadNodes}, #{self := Node}, _Config) ->
    ?LOG_INFO("Ecron Global UP on ~p at ~p quorum_size is ~p good_nodes is ~p bad_nodes is ~p.",
        [Node, Time, QuorumSize, GoodNodes, BadNodes]);
handle_event(?GlobalDown, #{action_ms := Time, quorum_size := QuorumSize,
    good_nodes := GoodNodes, bad_nodes := BadNodes}, #{self := Node}, _Config) ->
    ?LOG_INFO("Ecron Global DOWN on ~p at ~p quorum_size is ~p good_nodes is ~p bad_nodes is ~p.",
        [Node, Time, QuorumSize, GoodNodes, BadNodes]).
